notes/advanced-data-types.md

```diff
@@ -0,0 +1,530 @@
+T
+```
+/*
+enum class enumName {
+    constant1,
+    constant2,
+    ...
+    constantN
+};
+*/
+
+enum class betterWeekday {
+    betterSunday,
+    betterMonday,
+    betterTuesday,
+    betterWednesday,
+    betterThursday,
+    betterFriday,
+    betterSaturday
+};
+```
+<div style="text-align: center"> Figure 6a - Enum Class Declaration </div>
+
+To reference a constant within an `enum class`, you must use the scope resolution operator (`::`):
+
+<div style="text-align: center"> 
+
+`enumClassName::Field` 
+
+</div>
+
+This tells us that constants belonging to `enum class`es are specific to those classes, meaning that we can now define variables or other `enum class` constants with the same name. To demonstrate, let's draw an analogy to Figure 5a with the following valid example:
+
+```
+enum class People = {Good, Bad};
+
+int main() {
+    enum class Others = {Good, Meh}; // Okay now, since People::Good is not Others::Good
+    
+    int Meh = 0; // This is okay now, since main::Meh is not Others::Meh
+}
+
+enum Professors = {Okay, Meh} // Still okay
+```
+<div style="text-align: center"> Figure 7a - Demonstration of the Strongly Scoped Enum Class </div>
+
+However, this means the initialization of the `enum class` is different
+
+```
+// enumClassName enumClassVarName = enumClassName::Constant;
+
+betterWeekday betterDay = betterWeekday::betterWednesday;
+
+```
+<div style="text-align: center"> Figure 6b - Enum Class Initialization </div>
+
+With the use of the keyword `class`, C++ establishes `enum class`es as distinct entities, which does not allow for automatic type conversion.
+
+```
+// Pretend weekday and painMeter are now enum classes
+weekday day = weekday::Friday;
+painMeter painRating = painMeter::pain;
+if(painRating == painRating) {
+    std::cout << "Days are pain!!!" << std::endl;
+}
+
+// Output: Error, no suitible conversion from weekday to PainMeter elements
+```
+<div style="text-align: center"> Figure 7b - Demonstration of the Strongly Typed Enum Class </div>
+
+
+There is one more advantage you can control, and that is the underlying type of the constants in an `enum class`. We've accepted that it was an integer, but what if the constants could be less or more bytes? To specify the specific type you want your `enum class` constants to represent, simply use the following notation:
+```
+/*
+enum class enumClassName : type {
+    constant1,
+    constant2,
+    ...
+    constantN
+};
+*/
+
+enum class betterWeekday : char {
+    betterSunday,
+    betterMonday,
+    betterTuesday,
+    betterWednesday,
+    betterThursday,
+    betterFriday,
+    betterSaturday
+}; 
+
+```
+<div style="text-align: center"> Figure 8 - Enum Class Type Specification </div>
+
+## Unions (`union`)
+
+A union is a collection of different variables, like a struct, but it only holds one at a time. To do this, a union will always take up the amount of space specified by the largest variable type. To demonstrate, look at the following union:
```

Comment:
I think it's worth mentioning a few things:

1. Unions are used mainly to save memory, especially useful for embedded programming
2. All members of the union share the same memory location